# 📂 Loading Datasets

Real-world datasets are usually stored in files such as:

- CSV

- JSON

- Excel

Pandas can load these files into DataFrames for analysis.

## 📄 Loading CSV Files

### `read_csv()`

Used to load CSV (Comma Separated Values) files.

### Example


In [69]:
import pandas as pd
df = pd.read_csv("student_data.csv")
print(df)

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health 

## `to_string( )`

`to_string( )` converts a pandas DataFrame or Series into formatted text.

In [70]:
print(df.to_string())

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob      reason guardian  traveltime  studytime  failures schoolsup famsup paid activities nursery higher internet romantic  famrel  freetime  goout  Dalc  Walc  health  absences  G1  G2  G3
0       GP   F   18       U     GT3       A     4     4   at_home   teacher      course   mother           2          2         0       yes     no   no         no     yes    yes       no       no       4         3      4     1     1       3         6   5   6   6
1       GP   F   17       U     GT3       T     1     1   at_home     other      course   father           1          2         0        no    yes   no         no      no    yes      yes       no       5         3      3     1     1       3         4   5   5   6
2       GP   F   15       U     LE3       T     1     1   at_home     other       other   mother           1          2         3       yes     no  yes         no     yes    yes      yes       no       4        

## 🌐 Loading JSON Files

### `read_json()`

Used to load JSON datasets.

### Example

In [12]:
print(pd.read_json("student.json"))

   student_id           name  age  gender               department  semester  \
0       ST001       Ali Khan   20    Male         Computer Science         4   
1       ST002     Sara Ahmed   21  Female     Software Engineering         5   
2       ST003    Usman Tariq   22    Male  Artificial Intelligence         6   
3       ST004    Fatima Noor   20  Female             Data Science         4   
4       ST005      Hamza Ali   23    Male           Cyber Security         7   
5       ST006   Ayesha Malik   21  Female     Software Engineering         5   
6       ST007  Bilal Hussain   22    Male         Computer Science         6   
7       ST008      Hira Shah   20  Female  Artificial Intelligence         4   
8       ST009     Zain Ahmed   24    Male             Data Science         8   
9       ST010   Maham Yousaf   21  Female           Cyber Security         5   
10      ST011  Taha Siddiqui   22    Male     Software Engineering         6   
11      ST012    Noor Fatima   20  Femal

## 📘 Inspecting Real Datasets in Pandas

Before analyzing or cleaning data, we first inspect the dataset to understand:

- rows and columns
- datatypes
- missing values
- column names
- overall structure

This is one of the most important steps in Data Analysis and Machine Learning.

### Viewing First Rows with `head( )`

Displays the first 5 rows by default.

### Example


In [13]:
print(df.head())

  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]


***Custom Number of Rows***

In [14]:
print(df.head(8))

  school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4   at_home   teacher   
1     GP   F   17       U     GT3       T     1     1   at_home     other   
2     GP   F   15       U     LE3       T     1     1   at_home     other   
3     GP   F   15       U     GT3       T     4     2    health  services   
4     GP   F   16       U     GT3       T     3     3     other     other   
5     GP   M   16       U     LE3       T     4     3  services     other   
6     GP   M   16       U     LE3       T     2     2     other     other   
7     GP   F   17       U     GT3       A     4     4     other   teacher   

   ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0  ...      4        3      4     1     1      3        6   5   6   6  
1  ...      5        3      3     1     1      3        4   5   5   6  
2  ...      4        3      2     2     3      3       10   7   8  10  
3  ...      3     

### Viewing Last Rows with `tail( )`

Displays the last rows of the dataset.

### Example

In [15]:
print(df.tail())

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[5 rows x 33 columns]


### Dataset Shape with `shape`

Returns:

`(rows, columns)`

### Example

In [16]:
print(df.shape)

(395, 33)


### Column Names with `columns`

Displays all column names.

### Example

In [17]:
print(df.columns)

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='str')


### Data Types with `dtypes`

Shows datatype of each column.

### Example

In [18]:
print(df.dtypes)

school          str
sex             str
age           int64
address         str
famsize         str
Pstatus         str
Medu          int64
Fedu          int64
Mjob            str
Fjob            str
reason          str
guardian        str
traveltime    int64
studytime     int64
failures      int64
schoolsup       str
famsup          str
paid            str
activities      str
nursery         str
higher          str
internet        str
romantic        str
famrel        int64
freetime      int64
goout         int64
Dalc          int64
Walc          int64
health        int64
absences      int64
G1            int64
G2            int64
G3            int64
dtype: object


## Full Dataset Summary with `info( )`

One of the MOST used pandas functions.

Shows:

- total rows

- columns

- datatypes

- non-null values

- memory usage

### Example

In [19]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   school      395 non-null    str  
 1   sex         395 non-null    str  
 2   age         395 non-null    int64
 3   address     395 non-null    str  
 4   famsize     395 non-null    str  
 5   Pstatus     395 non-null    str  
 6   Medu        395 non-null    int64
 7   Fedu        395 non-null    int64
 8   Mjob        395 non-null    str  
 9   Fjob        395 non-null    str  
 10  reason      395 non-null    str  
 11  guardian    395 non-null    str  
 12  traveltime  395 non-null    int64
 13  studytime   395 non-null    int64
 14  failures    395 non-null    int64
 15  schoolsup   395 non-null    str  
 16  famsup      395 non-null    str  
 17  paid        395 non-null    str  
 18  activities  395 non-null    str  
 19  nursery     395 non-null    str  
 20  higher      395 non-null    str  
 21  inte

### Why Important?

Used to quickly detect:

- missing values
- wrong datatypes
- dataset structure





## Statistical Summary with `describe( )`

Provides statistics for numerical columns.

### Example

In [20]:
print(df.describe())

              age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.9988

## Statistical Summary of All Columns

Use include='all' inside describe function :

This also includes categorical/text columns.



In [21]:
print(df.describe(include="all"))

       school  sex         age address famsize Pstatus        Medu  \
count     395  395  395.000000     395     395     395  395.000000   
unique      2    2         NaN       2       2       2         NaN   
top        GP    F         NaN       U     GT3       T         NaN   
freq      349  208         NaN     307     281     354         NaN   
mean      NaN  NaN   16.696203     NaN     NaN     NaN    2.749367   
std       NaN  NaN    1.276043     NaN     NaN     NaN    1.094735   
min       NaN  NaN   15.000000     NaN     NaN     NaN    0.000000   
25%       NaN  NaN   16.000000     NaN     NaN     NaN    2.000000   
50%       NaN  NaN   17.000000     NaN     NaN     NaN    3.000000   
75%       NaN  NaN   18.000000     NaN     NaN     NaN    4.000000   
max       NaN  NaN   22.000000     NaN     NaN     NaN    4.000000   

              Fedu   Mjob   Fjob  ...      famrel    freetime       goout  \
count   395.000000    395    395  ...  395.000000  395.000000  395.000000   
uniqu

## Viewing Dataset Index

Displays row labels/indexes.

### Example

In [22]:
print(df.index)

RangeIndex(start=0, stop=395, step=1)


## column Selection & Filtering in Pandas

 column selection is one of the most important parts of Pandas.

It allows us to:

- access specific data

- filter useful information

- analyze datasets

- prepare data for Machine Learning

This is heavily used in:

- Data Analysis

- Exploratory Data Analysis (EDA)

- Data Cleaning

- Machine Learning preprocessing


### Selecting a Single Column

Selecting one column returns a **Series**.

### Example

In [23]:
print(df["school"])

0      GP
1      GP
2      GP
3      GP
4      GP
       ..
390    MS
391    MS
392    MS
393    MS
394    MS
Name: school, Length: 395, dtype: str


### Selecting Multiple Columns

Selecting multiple columns returns a **DataFrame**.

### Example


In [24]:
print(df[["school", "age"]])

    school  age
0       GP   18
1       GP   17
2       GP   15
3       GP   15
4       GP   16
..     ...  ...
390     MS   20
391     MS   17
392     MS   21
393     MS   18
394     MS   19

[395 rows x 2 columns]


## Filtering Data

Filtering is used to display only the rows that match a condition.

This is one of the MOST important Pandas skills.


### Single Condition Filtering


In [25]:
print(df[df["age"] > 15])

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
4       GP   F   16       U     GT3       T     3     3     other     other   
5       GP   M   16       U     LE3       T     4     3  services     other   
6       GP   M   16       U     LE3       T     2     2     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health 

### Multiple Conditions using `&` (AND)

`&` means BOTH conditions must be true.

### Example


In [26]:
guy_students = df[
    (df["sex"] == "M") & (df["age"] > 20)
]

print(guy_students)

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
247     GP   M   22       U     GT3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   

     ... famrel freetime  goout  Dalc  Walc health absences  G1 G2 G3  
247  ...      5        4      5     5     5      1       16   6  8  8  
392  ...      5        5      3     3     3      3        3  10  8  7  

[2 rows x 33 columns]



### Important

Each condition must be inside parentheses `( )`.


### Multiple Conditions using `|` (OR)

`|` means ANY ONE condition can be true.

### Example

In [27]:
mature = df[(df["sex"]=="M") | (df["age"] > 20)]
print(mature)

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
5       GP   M   16       U     LE3       T     4     3  services     other   
6       GP   M   16       U     LE3       T     2     2     other     other   
8       GP   M   15       U     LE3       A     3     2  services     other   
9       GP   M   15       U     GT3       T     3     4     other     other   
12      GP   M   15       U     LE3       T     4     4    health  services   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health 

## Sorting Data

### sort_values( )

Used to sort data based on a column.

### Example

In [28]:
print(df.sort_values("age", ascending=False))

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
247     GP   M   22       U     GT3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
390     MS   M   20       U     LE3       A     2     2  services  services   
376     MS   F   20       U     GT3       T     4     2    health     other   
306     GP   M   20       U     GT3       A     3     2  services     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
35      GP   F   15       U     GT3       T     2     3     other     other   
38      GP   F   15       R     GT3       T     3     4  services    health   
36      GP   M   15       U     LE3       T     4     3   teacher  services   
3       GP   F   15       U     GT3       T     4     2    health  services   
2       GP   F   15       U     LE3       T     1     1   at_home     other   

     ... famrel freetime  goout  Dalc  Walc health 

## Value Analysis

### value_counts( )

Counts frequency of each unique value in a column.

### Example

In [29]:
print(df["school"].value_counts())

school
GP    349
MS     46
Name: count, dtype: int64


### unique( )

Returns all unique values in a column.

### Example

In [30]:
print(df["Mjob"].unique())

<StringArray>
['at_home', 'health', 'other', 'services', 'teacher']
Length: 5, dtype: str


### nunique( )

Returns number of unique values in a column.

### Example

In [31]:
print(df["Fjob"].nunique())

5


## Grouping

### groupby( )

Groups data based on a column.

### Example

In [32]:
for name in df.groupby("school"):
    print(name)

('GP',     school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4  at_home   teacher   
1       GP   F   17       U     GT3       T     1     1  at_home     other   
2       GP   F   15       U     LE3       T     1     1  at_home     other   
3       GP   F   15       U     GT3       T     4     2   health  services   
4       GP   F   16       U     GT3       T     3     3    other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...      ...       ...   
344     GP   F   18       U     GT3       T     2     3  at_home     other   
345     GP   F   18       U     GT3       T     3     2    other  services   
346     GP   M   18       R     GT3       T     4     3  teacher  services   
347     GP   M   18       U     GT3       T     4     3  teacher     other   
348     GP   F   17       U     GT3       T     4     3   health     other   

     ... famrel freetime  goout  Dalc  Walc health absen

## Aggregation with groupby( )

- Splits data into groups

- Applies calculations per group

- Very important for real-world analysis



In [71]:
print(df.groupby("school").agg({"age":["mean","sum","max","min"]}))

              age              
             mean   sum max min
school                         
GP      16.521490  5766  22  15
MS      18.021739   829  21  17




- - -

# 🧩 Challenge: Real Dataset Mini EDA

## 🎯 Objective

Apply all Pandas concepts learned so far on a real dataset and perform basic exploratory data analysis (EDA).

You should be able to:

- load dataset

- explore structure

- filter data

- sort values

- analyze categories

- compute basic statistics

- extract insights

---


# 🧪 Tasks

## 1️⃣ Dataset Overview

- Display first 10 rows

- Check shape of dataset

- Show column names

- Check data types

- Generate summary using info()

---

## 2️⃣ Basic Statistics

- Show statistical summary using describe()

- Include categorical columns as well

---

## 3️⃣ Column Exploration

- Select any 3 columns

- Extract one column as Series

- Find number of unique values in a column

- Display value counts of a categorical column

---

## 4️⃣ Filtering Practice

- Filter rows based on one condition

- Apply at least one AND condition

- Apply at least one OR condition

Example ideas:

- age > 18

- gender == "male" AND survived == 1

- marks > 80 OR attendance > 90

---

## 5️⃣ Sorting Practice

- Sort dataset by one numeric column (ascending)

- Sort dataset by another column (descending)

---

## 6️⃣ Grouping Practice (CORE)
- Group data by a categorical column
- Find:
  - mean of numeric column
  - sum of numeric column
  - count per group

---

## 7️⃣ Final Insight (IMPORTANT)
Write 2–3 observations like:

- Which category is most frequent?
- Which group has highest average value?
- Any surprising pattern?

---

## 1️⃣ Dataset Overview

In [56]:
df = pd.read_csv("student_data.csv")
print(df.head(10))

  school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4   at_home   teacher   
1     GP   F   17       U     GT3       T     1     1   at_home     other   
2     GP   F   15       U     LE3       T     1     1   at_home     other   
3     GP   F   15       U     GT3       T     4     2    health  services   
4     GP   F   16       U     GT3       T     3     3     other     other   
5     GP   M   16       U     LE3       T     4     3  services     other   
6     GP   M   16       U     LE3       T     2     2     other     other   
7     GP   F   17       U     GT3       A     4     4     other   teacher   
8     GP   M   15       U     LE3       A     3     2  services     other   
9     GP   M   15       U     GT3       T     3     4     other     other   

   ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0  ...      4        3      4     1     1      3        6   5   6   6  
1  ...  

In [38]:
print(df.shape)

(395, 33)


In [39]:
print(df.columns)

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='str')


In [40]:
print(df.dtypes)

school          str
sex             str
age           int64
address         str
famsize         str
Pstatus         str
Medu          int64
Fedu          int64
Mjob            str
Fjob            str
reason          str
guardian        str
traveltime    int64
studytime     int64
failures      int64
schoolsup       str
famsup          str
paid            str
activities      str
nursery         str
higher          str
internet        str
romantic        str
famrel        int64
freetime      int64
goout         int64
Dalc          int64
Walc          int64
health        int64
absences      int64
G1            int64
G2            int64
G3            int64
dtype: object


In [43]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   school      395 non-null    str  
 1   sex         395 non-null    str  
 2   age         395 non-null    int64
 3   address     395 non-null    str  
 4   famsize     395 non-null    str  
 5   Pstatus     395 non-null    str  
 6   Medu        395 non-null    int64
 7   Fedu        395 non-null    int64
 8   Mjob        395 non-null    str  
 9   Fjob        395 non-null    str  
 10  reason      395 non-null    str  
 11  guardian    395 non-null    str  
 12  traveltime  395 non-null    int64
 13  studytime   395 non-null    int64
 14  failures    395 non-null    int64
 15  schoolsup   395 non-null    str  
 16  famsup      395 non-null    str  
 17  paid        395 non-null    str  
 18  activities  395 non-null    str  
 19  nursery     395 non-null    str  
 20  higher      395 non-null    str  
 21  inte

## 2️⃣ Basic Statistics

In [44]:
print(df.describe(include='all'))

       school  sex         age address famsize Pstatus        Medu  \
count     395  395  395.000000     395     395     395  395.000000   
unique      2    2         NaN       2       2       2         NaN   
top        GP    F         NaN       U     GT3       T         NaN   
freq      349  208         NaN     307     281     354         NaN   
mean      NaN  NaN   16.696203     NaN     NaN     NaN    2.749367   
std       NaN  NaN    1.276043     NaN     NaN     NaN    1.094735   
min       NaN  NaN   15.000000     NaN     NaN     NaN    0.000000   
25%       NaN  NaN   16.000000     NaN     NaN     NaN    2.000000   
50%       NaN  NaN   17.000000     NaN     NaN     NaN    3.000000   
75%       NaN  NaN   18.000000     NaN     NaN     NaN    4.000000   
max       NaN  NaN   22.000000     NaN     NaN     NaN    4.000000   

              Fedu   Mjob   Fjob  ...      famrel    freetime       goout  \
count   395.000000    395    395  ...  395.000000  395.000000  395.000000   
uniqu

## 3️⃣ Column Exploration

In [45]:
selected_columns = df[["school", "age", "sex"]]
print(selected_columns)

    school  age sex
0       GP   18   F
1       GP   17   F
2       GP   15   F
3       GP   15   F
4       GP   16   F
..     ...  ...  ..
390     MS   20   M
391     MS   17   M
392     MS   21   M
393     MS   18   M
394     MS   19   M

[395 rows x 3 columns]


In [54]:
series_column = df["age"]
print(series_column)

0      18
1      17
2      15
3      15
4      16
       ..
390    20
391    17
392    21
393    18
394    19
Name: age, Length: 395, dtype: int64


In [47]:
unique_schools = df["school"].unique()
print(unique_schools)

<StringArray>
['GP', 'MS']
Length: 2, dtype: str


In [50]:
value_counts = df["school"].value_counts()
print(value_counts)

school
GP    349
MS     46
Name: count, dtype: int64


## 4️⃣ Filtering Practice

In [51]:
filtered = df[((df["school"]=='GP') & (df["age"] > 15)) | ((df["school"]=='MS') & (df["age"] > 18))]
print(filtered)

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
4       GP   F   16       U     GT3       T     3     3     other     other   
5       GP   M   16       U     LE3       T     4     3  services     other   
6       GP   M   16       U     LE3       T     2     2     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
383     MS   M   19       R     GT3       T     1     1     other  services   
387     MS   F   19       R     GT3       T     2     3  services     other   
390     MS   M   20       U     LE3       A     2     2  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health 

## 5️⃣ Sorting Practice

In [58]:
ascending_sort = df.sort_values("age", ascending=True)
print(ascending_sort)


    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
21      GP   M   15       U     GT3       T     4     4    health    health   
20      GP   M   15       U     GT3       T     4     3   teacher     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
36      GP   M   15       U     LE3       T     4     3   teacher  services   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
376     MS   F   20       U     GT3       T     4     2    health     other   
306     GP   M   20       U     GT3       A     3     2  services     other   
390     MS   M   20       U     LE3       A     2     2  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
247     GP   M   22       U     GT3       T     3     1  services  services   

     ... famrel freetime  goout  Dalc  Walc health 

In [59]:
descending_sort = df.sort_values("age", ascending=False)
print(descending_sort)

    school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
247     GP   M   22       U     GT3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
390     MS   M   20       U     LE3       A     2     2  services  services   
376     MS   F   20       U     GT3       T     4     2    health     other   
306     GP   M   20       U     GT3       A     3     2  services     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
35      GP   F   15       U     GT3       T     2     3     other     other   
38      GP   F   15       R     GT3       T     3     4  services    health   
36      GP   M   15       U     LE3       T     4     3   teacher  services   
3       GP   F   15       U     GT3       T     4     2    health  services   
2       GP   F   15       U     LE3       T     1     1   at_home     other   

     ... famrel freetime  goout  Dalc  Walc health 

## 6️⃣ Grouping Practice

In [68]:
grouping = df.groupby("school").agg({"absences" : ["mean","sum","count"]})
print(grouping)

        absences            
            mean   sum count
school                      
GP      5.965616  2082   349
MS      3.760870   173    46


## 📊 Final Insights — Student Dataset Analysis

After performing EDA on the dataset, here are the key observations:

---

## 1️⃣ Most Frequent Category

- The most frequent school is:
 
  - **GP (349 students)**

- MS school has significantly fewer students:

  - **MS (46 students)**

👉 Conclusion:

The dataset is heavily dominated by GP students.

---

## 2️⃣ Age Analysis by School

| School | Average Age |
|--------|-------------|
| GP     | 16.52       |
| MS     | 18.02       |

👉 Conclusion:

- MS students are slightly older on average compared to GP students.

- This might indicate different academic levels or admission patterns.

---

## 3️⃣ Total Absences by School

| School | Total Absences |
|--------|----------------|
| GP     | 2082           |
| MS     | 173            |

👉 Conclusion:

- GP contributes most absences overall (likely due to more students).

- Absence pattern aligns with dataset size imbalance.

---

## 🧠 Key Takeaways

- Dataset is **imbalanced toward GP school**

- MS students are **older on average**

- Absences are mostly influenced by **group size**

- Basic grouping reveals **strong demographic differences**
